### PR TITLE
Implemented caching songs when played [#9]

### PR DESCRIPTION
**Pull request by taahamahdi**

### CODE REVIEW

File: helpers/utils.js [hunk 1]
```diff
@@ -1,7 +1,10 @@
 const RED = 0xE74C3C;
+const CACHE_DIR = "./.cache";
```
Comment (Brainicism) [inline helpers/utils.js:2]: Should specify the cache directory in `config.json.` Also name the variable something more descriptive, something like song_cache_dir

File: helpers/utils.js [hunk 2]
```diff
@@ -91,11 +94,46 @@ module.exports = {
...
+    if (!fs.existsSync(CACHE_DIR)) {
+        fs.mkdirSync(dir)
+    }
+
+    const ytdlOptions = {
+        filter: "audioonly",
+        quality: "highest"
+    };
+
+    const songLocation = `${CACHE_DIR}/${gameSession.getLink()}.mp3`;
```
Comment (Brainicism) [inline helpers/utils.js:116]: cachedSongLocation

Comment (Brainicism) [reply helpers/utils.js:116]: Another issue not related to this PR: 
`GameSession::getLink()` is misleading, it returns a youtube video ID, not the full URL. Could you refactor this?

File: helpers/utils.js [hunk 3]
```diff
@@ -91,11 +94,46 @@ module.exports = {
...
+    if (!fs.existsSync(songLocation)) {
+        gameSession.isSongCached = false;
+        const tempLocation = `${songLocation}.part`;
+        if (!fs.existsSync(tempLocation)) {
+            let cacheStream = fs.createWriteStream(tempLocation);
+            ytdl(gameSession.getLink(), ytdlOptions)
+                .pipe(cacheStream);
+            cacheStream.on('finish', () => {
+                fs.renameSync(tempLocation, songLocation)
```
Comment (Brainicism) [inline helpers/utils.js:125]: Can you create an issue to verify that the song was downloaded in full?

File: helpers/utils.js [hunk 4]
```diff
@@ -91,11 +94,46 @@ module.exports = {
...
+    }
+
+    const ytdlOptions = {
+        filter: "audioonly",
+        quality: "highest"
+    };
+
+    const songLocation = `${CACHE_DIR}/${gameSession.getLink()}.mp3`;
+    if (!fs.existsSync(songLocation)) {
+        gameSession.isSongCached = false;
```
Comment (Brainicism) [inline helpers/utils.js:118]: You can refactor this by doing `gameSession.isSongCached = !fs.existsSync(songLocation)` outside the `if`, then using that condition to check

File: helpers/utils.js [hunk 5]
```diff
@@ -91,11 +94,46 @@ module.exports = {
...
+        const tempLocation = `${songLocation}.part`;
+        if (!fs.existsSync(tempLocation)) {
```
Comment (Brainicism) [inline helpers/utils.js:120]: What happens if we have a partially completed download that was interrupted?

File: helpers/utils.js (hunk 2 above)
Comment (taahamahdi) [reply helpers/utils.js:116]: Sure and sure!

File: helpers/utils.js (hunk 5 above)
Comment (taahamahdi) [reply helpers/utils.js:120]: This needs to be fixed with that new issue you told me to create. It'll just stream the song instead, but the `.part` file will exist in an incomplete state.

**Pull request merged**